File: backend/app/services/alerts.py
"""Threshold-based alert service.

Evaluates CPU, memory, disk, monthly traffic, availability checks, service
state and the anomaly score on every collection. Alerts fire only after a
threshold stays breached for `alert_sustain_seconds` (no false alarms on
momentary spikes), are deduplicated and can be pushed to a webhook.
"""

import itertools
import time
from collections import deque
from datetime import datetime, timezone
from typing import Optional

from ..core.config import Settings
from .notifications import WebhookNotifier


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class AlertService:
    MAX_RESOLVED = 20

    def __init__(self, settings: Settings, logger=None, notifier: Optional[WebhookNotifier] = None):
        self.settings = settings
        self.logger = logger
        self.notifier = notifier
        self._active: dict[str, dict] = {}
        self._recent: deque[dict] = deque(maxlen=self.MAX_RESOLVED)
        self._pending: dict[str, dict] = {}
        self._seq = itertools.count(1)

    # --- Helpers -------------------------------------------------------------

    def _thresholds(self) -> dict:
        return {
            "cpu_warning": self.settings.alert_cpu_warning,
            "cpu_critical": self.settings.alert_cpu_critical,
            "memory_warning": self.settings.alert_memory_warning,
            "memory_critical": self.settings.alert_memory_critical,
            "disk_warning": self.settings.alert_disk_warning,
            "disk_critical": self.settings.alert_disk_critical,
            "traffic_warning": self.settings.traffic_warning_percent,
            "traffic_critical": self.settings.traffic_critical_percent,
            "anomaly_critical": self.settings.anomaly_critical,
        }

    def _notify(self, alert: dict) -> None:
        if self.notifier is not None and self.notifier.enabled:
            try:
                self.notifier.send(
                    {
                        "event": "alert_raised" if alert["state"] == "active" else "alert_resolved",
                        "severity": alert.get("severity"),
                        "title": alert.get("title"),
                        "message": alert.get("message"),
                        "tip": alert.get("tip"),
                        "metric": alert.get("metric"),
                        "value": alert.get("value"),
                        "threshold": alert.get("threshold"),
                    }
                )
            except Exception:  # noqa: BLE001 - notifications never crash alerting
                if self.logger:
                    self.logger.exception("notification failed")

    def _upsert(
        self,
        key: str,
        severity: str,
        title: str,
        message: str,
        tip: str,
        metric: Optional[str],
        value: Optional[float],
        threshold: Optional[float],
    ) -> None:
        now = _now_iso()
        alert = self._active.get(key)
        if alert is not None:
            alert["last_seen"] = now
            if severity == "critical" and alert["severity"] != "critical":
                alert["severity"] = severity
                alert["message"] = message
                alert["value"] = value
                alert["threshold"] = threshold
                self._notify(alert)
            return
        self._active[key] = {
            "id": f"al-{next(self._seq)}",
            "key": key,
            "severity": severity,
            "title": title,
            "message": message,
            "tip": tip,
            "state": "active",
            "metric": metric,
            "value": value,
            "threshold": threshold,
            "first_seen": now,
            "last_seen": now,
            "resolved_at": None,
        }
        if self.logger:
            self.logger.warning("alert raised", extra={"event": key, "severity": severity})
        self._notify(self._active[key])

    def _resolve(self, key: str) -> None:
        alert = self._active.pop(key, None)
        if alert is not None:
            alert["state"] = "resolved"
            alert["resolved_at"] = _now_iso()
            self._recent.appendleft(alert)
            if self.logger:
                self.logger.info("alert resolved", extra={"event": key})
            self._notify(alert)

    def _check_threshold(
        self,
        key: str,
        label: str,
        value: float,
        warning: float,
        critical: float,
        tip: str,
        sustain: bool = True,
    ) -> None:
        """Raises an alert when the value stays above the threshold for the
        configured duration; resolves it once the value drops below warning."""
        metric = key.split("_")[0]
        if value >= warning:
            severity = "critical" if value >= critical else "warning"
            pending = self._pending.get(key)
            if pending is None:
                self._pending[key] = {"since": time.monotonic(), "severity": severity}
            elif severity == "critical" and pending["severity"] != "critical":
                self._pending[key] = {"since": time.monotonic(), "severity": "critical"}
            else:
                pending["severity"] = severity
            sustained = not sustain or time.monotonic() - self._pending[key]["since"] >= self.settings.alert_sustain_seconds
            if sustained and key not in self._active:
                self._upsert(
                    key, severity, f"{label} {'crítico' if severity == 'critical' else 'alto'}",
                    (
                        f"El uso de {label.lower()} alcanzó {value:.0f}% (umbral crítico {critical:.0f}%)."
                        if severity == "critical"
                        else f"El uso de {label.lower()} está en {value:.0f}% (umbral de aviso {warning:.0f}%)."
                    ),
                    tip, metric=metric, value=value, threshold=critical if severity == "critical" else warning,
                )
            return
        self._pending.pop(key, None)
        self._resolve(key)

    # --- Public API -----------------------------------------------------------

    def evaluate(
        self,
        summary: Optional[dict] = None,
        services=None,
        checks: Optional[list[dict]] = None,
    ) -> list[dict]:
        """Evaluates alerts for the given summary, services and/or checks."""
        if summary is not None:
            cpu = (summary.get("cpu") or {}).get("percent", 0)
            memory = (summary.get("memory") or {}).get("percent", 0)
            disks = summary.get("disks") or []
            disk = max(disks, key=lambda d: d.get("percent", 0), default=None)
            disk_percent = disk.get("percent", 0) if disk else 0

            self._check_threshold(
                "cpu_high", "CPU", cpu,
                self.settings.alert_cpu_warning, self.settings.alert_cpu_critical,
                "Revisa la vista de Procesos para identificar qué proceso consume más CPU.",
            )
            self._check_threshold(
                "memory_high", "Memoria", memory,
                self.settings.alert_memory_warning, self.settings.alert_memory_critical,
                "Considera liberar memoria o detener aplicaciones pesadas.",
            )
            mountpoint = disk.get("mountpoint", "/") if disk else "/"
            self._check_threshold(
                "disk_full", "Disco", disk_percent,
                self.settings.alert_disk_warning, self.settings.alert_disk_critical,
                f"Libera espacio en la partición {mountpoint} (archivos temporales, logs, paquetes sin usar).",
            )

            traffic = summary.get("traffic")
            if traffic and traffic.get("percent") is not None:
                self._check_threshold(
                    "traffic_quota", "Tráfico mensual", traffic["percent"],
                    self.settings.traffic_warning_percent, self.settings.traffic_critical_percent,
                    "Consulta el uso por servicio o revisa backups y transferencias programadas.",
                )

            anomaly = summary.get("anomaly")
            if anomaly:
                score = anomaly.get("score", 0)
                self._check_threshold(
                    "anomaly_behavior", "Comportamiento del sistema", score,
                    self.settings.anomaly_critical, self.settings.anomaly_critical,
                    "Revisa los procesos y el tráfico reciente: algo se desvía del comportamiento habitual.",
                )

        if services is not None:
            for item in getattr(services, "tracked", []):
                key = f"service_down_{item.name}"
                if item.state == "failed":
                    self._upsert(
                        key, "critical", f"Servicio {item.name} caído",
                        f"El servicio {item.name} está en estado fallido y no responde.",
                        f"Revisa los logs con: journalctl -u {item.name}.service",
                        metric="service", value=None, threshold=None,
                    )
                else:
                    self._resolve(key)

        if checks is not None:
            for check in checks:
                key = f"check_down_{check['name']}"
                if check.get("state") == "down":
                    self._upsert(
                        key, "critical", f"Verificación {check['name']} sin respuesta",
                        f"El objetivo {check['target']} no responde ({check.get('error') or 'error de conexión'}).",
                        "Comprueba el proceso, el puerto y el firewall del objetivo.",
                        metric="check", value=None, threshold=None,
                    )
                else:
                    self._resolve(key)
        return list(self._active.values())

    def snapshot(self) -> dict:
        active = sorted(self._active.values(), key=lambda a: a["last_seen"], reverse=True)
        recent = list(self._recent)
        return {
            "active_count": len(active),
            "thresholds": self._thresholds(),
            "alerts": active + recent,
        }

File: backend-node/src/alerts/alerts.ts
import type { Settings } from '../config/settings';
import { NotificationHub } from '../notifications/notifications';

interface Alert {
  id: string;
  key: string;
  severity: string;
  title: string;
  message: string;
  tip: string;
  state: 'active' | 'resolved';
  metric: string | null;
  value: number | null;
  threshold: number | null;
  first_seen: string;
  last_seen: string;
  resolved_at: string | null;
}

interface ThresholdState {
  since: number;
  severity: string;
}

export class AlertService {
  private readonly active: Record<string, Alert> = {};
  private readonly recent: Alert[] = [];
  private readonly pending: Record<string, ThresholdState> = {};
  private seq = 0;

  constructor(
    private readonly settings: Settings,
    private readonly logger: { warn(...args: unknown[]): void; info(...args: unknown[]): void; exception(...args: unknown[]): void },
    private readonly notifier?: { enabled: boolean; send(event: unknown): Promise<boolean> | boolean },
  ) {}

  thresholds(): Record<string, number> {
    const s = this.settings;
    return {
      cpu_warning: s.alertCpuWarning,
      cpu_critical: s.alertCpuCritical,
      memory_warning: s.alertMemoryWarning,
      memory_critical: s.alertMemoryCritical,
      disk_warning: s.alertDiskWarning,
      disk_critical: s.alertDiskCritical,
      traffic_warning: s.trafficWarningPercent,
      traffic_critical: s.trafficCriticalPercent,
      anomaly_critical: s.anomalyCritical,
    };
  }

  async notify(alert: Alert): Promise<void> {
    if (!this.notifier || !this.notifier.enabled) return;
    try {
      await this.notifier.send({
        event: alert.state === 'active' ? 'alert_raised' : 'alert_resolved',
        severity: alert.severity,
        title: alert.title,
        message: alert.message,
        tip: alert.tip,
        metric: alert.metric,
        value: alert.value,
        threshold: alert.threshold,
      });
    } catch {
      this.logger.exception?.('notification failed');
    }
  }

  upsert(
    key: string,
    severity: string,
    title: string,
    message: string,
    tip: string,
    metric: string | null,
    value: number | null,
    threshold: number | null,
  ): void {
    const now = new Date().toISOString();
    const alert = this.active[key];
    if (alert) {
      alert.last_seen = now;
      if (severity === 'critical' && alert.severity !== 'critical') {
        alert.severity = severity;
        alert.message = message;
        alert.value = value;
        alert.threshold = threshold;
        void this.notify(alert);
      }
      return;
    }
    this.seq += 1;
    const next: Alert = {
      id: `al-${this.seq}`,
      key,
      severity,
      title,
      message,
      tip,
      state: 'active',
      metric,
      value,
      threshold,
      first_seen: now,
      last_seen: now,
      resolved_at: null,
    };
    this.active[key] = next;
    this.logger.warn?.('alert raised', { event: key, severity });
    void this.notify(next);
  }

  private resolve(key: string): void {
    const alert = this.active[key];
    if (!alert) return;
    delete this.active[key];
    alert.state = 'resolved';
    alert.resolved_at = new Date().toISOString();
    this.recent.unshift(alert);
    if (this.recent.length > AlertService.MAX_RESOLVED) {
      this.recent.length = AlertService.MAX_RESOLVED;
    }
    this.logger.info?.('alert resolved', { event: key });
    void this.notify(alert);
  }

  checkThreshold(
    key: string,
    label: string,
    value: number,
    warning: number,
    critical: number,
    tip: string,
    sustain = true,
  ): void {
    const metric = key.split('_')[0];
    if (value >= warning) {
      const severity = value >= critical ? 'critical' : 'warning';
      const pending = this.pending[key];
      if (!pending) {
        this.pending[key] = { since: Date.now() / 1000, severity };
      } else if (severity === 'critical' && pending.severity !== 'critical') {
        this.pending[key] = { since: Date.now() / 1000, severity: 'critical' };
      } else {
        pending.severity = severity;
      }
      const sustainedSeconds = Date.now() / 1000 - this.pending[key].since;
      const sustained = !sustain || sustainedSeconds >= this.settings.alertSustainSeconds;
      if (sustained && !this.active[key]) {
        const isCritical = severity === 'critical';
        this.upsert(
          key,
          severity,
          `${label} ${isCritical ? 'crítico' : 'alto'}`,
          isCritical
            ? `El uso de ${label.toLowerCase()} alcanzó ${Math.round(value)}% (umbral crítico ${Math.round(critical)}%).`
            : `El uso de ${label.toLowerCase()} está en ${Math.round(value)}% (umbral de aviso ${Math.round(warning)}%).`,
          tip,
          metric,
          value,
          isCritical ? critical : warning,
        );
      }
      return;
    }
    delete this.pending[key];
    this.resolve(key);
  }

  async evaluate(
    summary: Record<string, unknown> | null,
    services: { tracked: Array<{ name: string; state: string }> } | null,
    checks: Array<{ name: string; state: string; target: string; error?: string | null }> | null,
  ): Promise<unknown[]> {
    if (summary) {
      const cpu = (summary['cpu'] ?? {}) as Record<string, unknown>;
      const memory = (summary['memory'] ?? {}) as Record<string, unknown>;
      const disks = (summary['disks'] ?? []) as Array<{ percent?: number; mountpoint?: string }>;
      const disk = disks.length
        ? disks.reduce((max, d) => ((d.percent ?? 0) > (max.percent ?? 0) ? d : max), disks[0])
        : null;
      const diskPercent = disk?.percent ?? 0;

      this.checkThreshold(
        'cpu_high', 'CPU', Number(cpu['percent'] ?? 0),
        this.settings.alertCpuWarning, this.settings.alertCpuCritical,
        'Revisa la vista de Procesos para identificar qué proceso consume más CPU.',
      );
      this.checkThreshold(
        'memory_high', 'Memoria', Number(memory['percent'] ?? 0),
        this.settings.alertMemoryWarning, this.settings.alertMemoryCritical,
        'Considera liberar memoria o detener aplicaciones pesadas.',
      );
      const mountpoint = disk?.mountpoint ?? '/';
      this.checkThreshold(
        'disk_full', 'Disco', diskPercent,
        this.settings.alertDiskWarning, this.settings.alertDiskCritical,
        `Libera espacio en la partición ${mountpoint} (archivos temporales, logs, paquetes sin usar).`,
      );

      const trafficResult = (summary['traffic'] ?? null) as { percent?: number | null } | null;
      if (trafficResult?.percent != null) {
        this.checkThreshold(
          'traffic_quota', 'Tráfico mensual', trafficResult.percent,
          this.settings.trafficWarningPercent, this.settings.trafficCriticalPercent,
          'Consulta el uso por servicio o revisa backups y transferencias programadas.',
        );
      }

      const anomaly = (summary['anomaly'] ?? null) as { score?: number } | null;
      if (anomaly) {
        this.checkThreshold(
          'anomaly_behavior', 'Comportamiento del sistema', Number(anomaly.score ?? 0),
          this.settings.anomalyCritical, this.settings.anomalyCritical,
          'Revisa los procesos y el tráfico reciente: algo se desvía del comportamiento habitual.',
        );
      }
    }

    if (services) {
      for (const item of services.tracked) {
        const key = `service_down_${item.name}`;
        if (item.state === 'failed') {
          this.upsert(
            key, 'critical', `Servicio ${item.name} caído`,
            `El servicio ${item.name} está en estado fallido y no responde.`,
            `Revisa los logs con: journalctl -u ${item.name}.service`,
            'service', null, null,
          );
        } else {
          this.resolve(key);
        }
      }
    }

    if (checks) {
      for (const check of checks) {
        const key = `check_down_${check.name}`;
        if (check.state === 'down') {
          this.upsert(
            key, 'critical', `Verificación ${check.name} sin respuesta`,
            `El objetivo ${check.target} no responde (${check.error ?? 'error de conexión'}).`,
            'Comprueba el proceso, el puerto y el firewall del objetivo.',
            'check', null, null,
          );
        } else {
          this.resolve(key);
        }
      }
    }
    return Object.values(this.active);
  }

  static readonly MAX_RESOLVED = 20;

  snapshot(): Record<string, unknown> {
    const active = Object.values(this.active).sort((a, b) => {
      return new Date(b.last_seen).getTime() - new Date(a.last_seen).getTime();
    });
    return {
      active_count: active.length,
      thresholds: this.thresholds(),
      alerts: [...active, ...this.recent],
    };
  }
}